### ISSUE TITLE
Changing index order of a DisplayObject after drawn to stage

### ISSUE THREAD

**Issue by iamdustan** [uxebu/bonsai#123]
In attempting to send a DisplayObject `toBack`, I couldn’t find any way to change the index order of an element after it’s already drawn to stage.

I image the following methods being useful for `DisplayObject`’s:
- `toIndex(i)` => moves to that index point in whatever it is drawn in (stage or group) 
- `moveBefore(DisplayObject)` => sugar around [`remove`](https://github.com/uxebu/bonsai/blob/master/src/runner/display_object.js#L708-723) and [`addBefore`](https://github.com/uxebu/bonsai/blob/master/src/runner/display_object.js#L692-706)
- `moveAfter(DisplayObject)` => same as above but with [`addAfter`](https://github.com/uxebu/bonsai/blob/master/src/runner/display_object.js#L676-690)
- `toFront()` => move to top (sugar around `remove` and `addTo`?)
- `toBack()` => move to bottom

Thoughts?

**Comment by padolsey** [comment]
Sorry for the late reply. Have you tried? -

``` js
// Remove and re-add at specified index:
fooDisplayObject.remove().addTo( theParent, theIndex );
```

**Comment by padolsey** [comment]
I agree that it would be nice to offer a couple of convenience methods though.

**Comment by iamdustan** [comment]
No worries. I have not tried playing with it at all yet. Been stuck in a land of matrix transformations instead. Will be looking more in-depth into this soon and will have more feedback. I was expecting to do that code example you provided for the `toIndex(i)` method.

**Comment by iamdustan** [comment]
So my initial take at this looks like the following:

``` javascript
DisplayObject.prototype.toIndex = function(index) {
  var parent = this.parent;
  this.remove().addTo(parent, index);
  return this;
};
```

This is working just fine with one caveat. Since we’re merely moving it, it would be nice to capture the `removedFromStage` event (https://github.com/uxebu/bonsai/blob/master/src/runner/display_object.js#L497). Additionally to capture the `addedToStage` event should also be avoided.

Thoughts?

**Comment by basecode** [comment]
As long as no "change index" event is offered I think it's okay to fire `removeFromStage` and `addedToStage`.

Since there's no need to cache the parent property and `addTo` already returns the DisplayObject instance, you could also write:

```
DisplayObject.prototype.toIndex = function(index) {
  return this.remove().addTo(this.parent, index);
};
```

Have you considered moving these wrappers in a plugin? See for example: https://github.com/bonsai-plugins/sprited-bitmap/blob/master/src/sprited_bitmap.js

**Comment by iamdustan** [comment]
A `changeIndex` event would be slick. I feel like the `addedToStage` and `removedFromStage` event in this scenario is a false positive when doing this.

I’m writing the index moving functions as a plugin in my current project to explore them. Do you think it would fit in core or be better left as a plugin?

**Comment by davidaurelio** [comment]
I think it will go into core. I agree on the events, and that should not be too tricky. `changeIndex`, `indexChange`, `changeDepth`, `depthChange`, `move`? `move` seems a little too generic

**Comment by iamdustan** [comment]
I vote for `changeIndex` for the following reasons:
- other events in bonsai follow the verb-->noun structure. `addedToStage` `removedFromStage` etc.
- `Index` keeps the same verbatim when referring to depth. This may be a biased decision towards those who work in the DOM, but seems inline with the current API such as `getIndexOfChild()`

**Comment by davidaurelio** [comment]
Ok, we have a winner.

**Comment by iamdustan** [comment]
@basecode fwiw I tried the shortened version you have and it fails at https://github.com/uxebu/bonsai/blob/master/src/runner/display_object.js#L718 because `parent === undefined`

**Comment by basecode** [comment]
Oh, of course. Please ignore my "pseudo-optimisation".

**Comment by iamdustan** [comment]
Here’s a start in what could be used as a plugin right now. https://gist.github.com/4161894

I see some issues with how it could work with if you try to add it to a different Group or DisplayList. Or if the remove changes the index there may need to be some checking to say if this was before do this or if this was after do that.

**Comment by davidaurelio** [comment]
Thanks, I will look into it in the next days
